File: solutions/TypeScript/[0921_Medium] Minimum Add to Make Parentheses Valid/minAddToMakeValid.ts
function minAddToMakeValid(s: string): number {
  let count = 0;
  const buffer: string[] = [];
  for (let i = 0; i < s.length; i++) {
    if (s[i] === ')') {
      if (buffer[buffer.length - 1] !== '(') {
        count++;
      }

      buffer.pop();
    } else {
      buffer.push(s[i]);
    }
  }

  if (buffer.length) {
    count += buffer.length;
  }

  return count;
}

File: solutions/Java/[0921_Medium] Minimum Add to Make Parentheses Valid/minAddToMakeValid.java
class Solution {
  public int minAddToMakeValid(String s) {
    int count = 0;
    Stack<Character> stack = new Stack<>();

    for (int i = 0; i < s.length(); i++) {
      if (s.charAt(i) == ')') {
        if (stack.empty() || stack.peek() != '(') {
          count++;
        } else {
          stack.pop();
        }
      } else {
        stack.push('(');
      }
    }

    count += stack.size();

    return count;
  }
}
